### PR TITLE
os/bluestore: fix buffers pinned by indefinitely deferred writes

### PR DESCRIPTION
**Pull request by liewegas**
Two problems:

- We could queue a deferred write and then never submit it if enough
other deferred writes didn't come along.  This would pin TransContexts
in the OpSequencer queue indefinitely.

- TransContexts' IOContext have a running_aios list with aio_t's, each
with a bufferlist pinning the buffers under IO.  Even after the IO
was completed, this list wasn't cleared, which meant buffers were
pinned as long as the TransContext stuck around.

The combination of these two things could make memory usage balloon
under certain workloads!

### CODE REVIEW

**Comment by markhpc** [comment]
So far this is working beautifully.  After 30 minutes of 4k random writes with 4k min alloc size the OSD is using approx 1.5GB RSS while previously it would be using 6GB+ RSS.

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -7815,6 +7816,10 @@ void BlueStore::_txc_finish(TransContext *txc)
 	  // for _osr_drain_preceding()
           notify = true;
 	}
+	if (txc->state == TransContext::STATE_DEFERRED_QUEUED &&
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: what's about an alternative approach to submit deferred txcs if deferred queue hasn't been updated for some time?

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -7815,6 +7816,10 @@ void BlueStore::_txc_finish(TransContext *txc)
...
+	    osr->q.size() > 32) {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: make a symbolic const or a config parameter?

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-06-02_08:32:01-rados-wip-sage-testing-distro-basic-smithi/

**Pull request merged**